### PR TITLE
[Bugfix][Determinism] Fix the batch-invariant softmax override

### PR DESCRIPTION
**Pull request by LH-and-FPGA**
# [Bugfix][Determinism] Fix batch-invariant softmax override

## Summary

Fixes two issues in `softmax_batch_invariant`, used when `VLLM_BATCH_INVARIANT=1`:

1. **The implementation was not actually batch-invariant.**  
   It used `torch.sum(exp_x, dim=...)`, whose reduction strategy depends on tensor shape. On RTX 5090 / fp32 / vocab=151936, the same row differed by up to 2 ULP between batch sizes.

2. **`half_to_float` was ignored.**  
   `aten::_softmax` has schema `(Tensor self, int dim, bool half_to_float)`, but the override treated the third argument as `dtype=None`. As a result, fp16 `softmax(..., dtype=torch.float32)` incorrectly returned fp16.

This affects all `torch.softmax` calls while batch-invariant mode is enabled, including sampling and MoE routing.

## Fix

- Reuse the existing row-per-program Triton log-softmax kernel through a `LOG: tl.constexpr` switch.
- Add `softmax()` alongside `log_softmax()`.
- Mirror `_log_softmax_batch_invariant` handling of `half_to_float`.
- Support non-last dimensions via `movedim` and handle empty tensors.
- Remove the dead `aten::softmax` registration (`aten::softmax` has no default overload).

One Triton program processes one row, so each row's reduction order is independent of batch size.

## Related PRs

- **#52231** contains a similar last-dim implementation as part of ROCm support, but does not fix `half_to_float`, interior dimensions, or the dead registration. This PR is a standalone CUDA-side bugfix and can land independently.
- **#49367** optimizes the log-softmax Triton kernel using split reduction. It does not modify `softmax_batch_invariant`; whichever PR lands second may need a small rebase.

No open issue currently reports this bug.

## Tests

RTX 5090 (sm_120), PyTorch 2.13.0+cu130, Triton 3.7.1.

```text
Before: 3 failed, 14 passed
After:  19 passed
```

Previously failing cases:
- fp16 `half_to_float`
- fp32 batch invariance at vocab sizes 4097 and 151936

Dispatcher-level validation:

```text
fp16 softmax(dtype=float32) -> float32
bf16 softmax(dtype=float32) -> float32
B=1/2/7/32/64/255 vs B=256: bitwise equal
log_softmax B=1 vs B=256:   bitwise equal
```

E2E batch-invariance tests with Qwen3-0.6B:

```text
MRv1 native sampler: 4 passed
MRv2:                2 passed
```

`pre-commit` also passes.

## Performance

fp32, vocab=151936, ms/call:

| batch | old override | native | this PR |
| ----: | -----------: | -----: | ------: |
| 1     | 0.021 | 0.035 | 0.146 |
| 8     | 0.025 | 0.037 | 0.150 |
| 64    | 0.105 | 0.038 | 0.152 |
| 256   | 0.771 | 0.347 | 0.422 |

The row-per-program kernel is slower at small batch sizes, matching the existing log-softmax behavior. #49367's split-reduction optimization could also benefit this kernel.

## Model impact

The change only affects `VLLM_BATCH_INVARIANT=1`.

Numerical differences from native `torch.softmax` remain within fp32 rounding (`max abs diff <= 4.2e-07`), while seeded E2E tests confirm tokens and logprobs are bitwise identical across batch sizes.

## AI assistance

Developed with AI assistance (Claude Code). All changes were reviewed by the submitter and the tests above were run locally.

### CODE REVIEW

**Review by Manny7717 — COMMENTED** [review]
Independent review — the two dispatch/signature claims verified by execution on this host (torch 2.13, CPU); the Triton kernel itself cannot be executed here (the 19 tests skip on CPU), so this is a logic+dispatch review rather than a full execution APPROVE.

**Claim 1 (half_to_float was ignored) — CONFIRMED.** Overrode `aten::_softmax` on CPU and traced every user-level softmax call: `torch.softmax`, `torch.nn.functional.softmax` (with and without `dtype=`), and `torch._softmax` all funnel into `aten::_softmax(Tensor self, int dim, bool half_to_float)`. The old override signature `softmax_batch_invariant(input, dim, dtype=None)` bound the third positional arg — a bool on CUDA — to `dtype` and never consulted it, so `torch._softmax(fp16_tensor, dim, True)` (fp32 result expected) went through the exp/sum path in fp16 and returned fp16. The new `if half_to_float: return softmax(input.float(), dim=dim)` matches the schema contract (upcast → fp32 out). Direct probe: `torch._softmax(x_fp16, -1, True)` on unpatched semantics returns fp16; expected fp32.

**Claim 2 (dead aten::softmax registration) — CONFIRMED.** Schema lookup on torch 2.13: `aten::softmax` default overload does not exist (only `aten::softmax.int` and `.Dimname` composites, which delegate to `_softmax`); `aten::_softmax` default exists. The removed registration could never dispatch; keeping only `_softmax` is correct since that is the op every softmax call path reaches (verified by the trace above).

**Non-last-dim + empty-tensor logic — verified by emulation.** Mirrored `_row_softmax`'s movedim recursion + reshape in plain torch on CPU: `softmax(x, dim=1)` over (4,5,6) fp32 matches `torch.nn.functional.softmax` to 0.0 maxdiff; empty (0,5) input returns empty_like without launch. The `dim<0` normalization and the movedim-back + `.contiguous()` round-trip are shape-safe.

**Kernel structure:** reuse of the pre-existing, production-tested row-per-program log-softmax kernel via a `LOG: tl.constexpr` switch is minimal-risk, and the row-per-program design is what actually delivers batch invariance (each row's reduction order is independent of row count), fixing the shape-dependent `torch.sum` reduction.

One non-blocking nit: in the non-LOG branch, `exp_vals = tl.exp(vals - max_val)` is computed in the input dtype's arithmetic, so for fp16/bf16 inputs *without* half_to_float the exp and the `sum_exp` accumulation stay at reduced precision (torch's CUDA softmax accumulates in fp32 even when returning fp16). This matches the pre-existing log-softmax path's behavior (same code shape, in production), and the half_to_float=True path correctly upcasts first — flagging only so it is a conscious choice. If fp16 half_to_float=False precision matters for sampling, upcasting unconditionally (fp32 compute → cast down) would match torch more closely at a modest cost.

CI: pre-run-check failure is the standard first-time-contributor gate (DCO passes), not a code issue.

**Comment by yuvalluria** [comment]
Tested on H100 NVL (SM 9.0), vLLM nightly (`0.28.1rc1.dev337+g27a94d1ce`), Torch 2.13.0+cu130.

Applied the PR's `batch_invariant.py` to the nightly image and ran `test_softmax_batch_invariant.py`.

```
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-9.1.1, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /workspace/testdir-pr55249
plugins: anyio-4.15.0
collecting ... collected 19 items

test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype0-1000] PASSED [  5%]
test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype0-4096] PASSED [ 10%]
test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype0-151936] PASSED [ 15%]
test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype1-1000] PASSED [ 21%]
test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype1-4096] PASSED [ 26%]
test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype1-151936] PASSED [ 31%]
test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype2-1000] PASSED [ 36%]
test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype2-4096] PASSED [ 42%]
test_softmax_batch_invariant.py::test_softmax_matches_torch[dtype2-151936] PASSED [ 47%]
test_softmax_batch_invariant.py::test_softmax_half_to_float PASSED       [ 52%]
test_softmax_batch_invariant.py::test_softmax_batch_invariant[dtype0-4097] PASSED [ 57%]
test_softmax_batch_invariant.py::test_softmax_batch_invariant[dtype0-151936] PASSED [ 63%]
test_softmax_batch_invariant.py::test_softmax_batch_invariant[dtype1-4097] PASSED [ 68%]
test_softmax_batch_invariant.py::test_softmax_batch_invariant[dtype1-151936] PASSED [ 73%]
test_softmax_batch_invariant.py::test_softmax_non_last_dim[0] PASSED     [ 78%]
test_softmax_batch_invariant.py::test_softmax_non_last_dim[1] PASSED     [ 84%]
test_softmax_batch_invariant.py::test_softmax_non_last_dim[-2] PASSED    [ 89%]
test_softmax_batch_invariant.py::test_softmax_empty[shape0] PASSED       [ 94%]
test_softmax_batch_invariant.py::test_softmax_empty[shape1] PASSED       [100%]

============================== 19 passed in 4.58s ==============================
```
